### DISCUSSION TITLE
Environment variables in hooks not used correctly?

### DISCUSSION THREAD

**Discussion by fabiangeisler** [cpenv/cpenv#60] [Q&A] [answered]
Not sure if I hit a bug or if this is intended behaviour. 
Consider the following setup:
- Module A has a `pre_activate.py` hook that sets env variable "FOO" (in python with `os.environ`)
- Module B has an environment in his `module.yml` defined that uses "FOO"
  like:
  ```yaml
  environment: 
    BAR: '${FOO}/somewhere'
  ```
- Module A is loaded before Module B in a test environment C

I would expect that "BAR" in Module B resolves fine with the variable from Module A, but it does not. Is this intended or not?

Side note:
That it is not correctly displayed in the `tk-cpenv` GUI is ok for me as long as the final resolved environment is working fine.

The behaviour that I am after is, setting an environment variable dynamically based on some conditions. Is there another/better way to achieve this in `cpenv`?

**Comment by danbradham** [comment, marked as answer]
I never intended to have environment variables that are set in hooks to be usable as variables in module files. It does make sense to expect that though so I'm labelling this as a bug.

The pre_activate hook is the best place to do those conditional variable assignments atm.

I'm not sure this is an issue we could resolve using the pre_activate hook since that runs specifically just before activation. Maybe there could be a new hook that runs just before environment variables are resolved and combined.

**Reply by fabiangeisler** [reply]
Maybe it is also possible to make the environment property on the modules settable? 
So you would be able to manipulate the environment variables on the live python Module instance, which than is used for resolving. I don't know though, if that is ok with your development intentions for the Module class. To me it would make sense since it gets passed into the hook anyway. Like, if you give me the object, why not let me manipulate it? 🙂
